### PR TITLE
GH-428 add client ssl context to suppor secure endpoint proxying

### PR DESCRIPTION
**Pull request by JacobJohansen**

### CODE REVIEW

File: zuul-core/src/main/java/com/netflix/zuul/netty/connectionpool/DefaultOriginChannelInitializer.java [hunk 1]
```diff
@@ -82,6 +93,14 @@ protected void initChannel(Channel ch) throws Exception {
     protected void addMethodBindingHandler(ChannelPipeline pipeline) {
     }
 
+    private SslContext getClientSslContext() throws SSLException {
+        return SslContextBuilder
+                    .forClient()
+                    .sslProvider(BaseSslContextFactory.chooseSslProvider())
+                    .startTls(true)
```
Comment (artgon) [inline zuul-core/src/main/java/com/netflix/zuul/netty/connectionpool/DefaultOriginChannelInitializer.java]: Why enable the `startTls` flag?

Comment (JacobJohansen) [reply zuul-core/src/main/java/com/netflix/zuul/netty/connectionpool/DefaultOriginChannelInitializer.java]: this triggers the ssl handshake for the connection

File: zuul-core/src/main/java/com/netflix/zuul/netty/connectionpool/DefaultOriginChannelInitializer.java [hunk 2]
```diff
@@ -82,6 +93,14 @@ protected void initChannel(Channel ch) throws Exception {
...
+                    .build();
```
[the remark this answers is not in the document]
Comment (artgon) [reply zuul-core/src/main/java/com/netflix/zuul/netty/connectionpool/DefaultOriginChannelInitializer.java]: Probably worth configuring the `.ciphers()` explicitly as well. See: https://github.com/Netflix/zuul/blob/2.1/zuul-core/src/main/java/com/netflix/netty/common/ssl/ServerSslConfig.java#L115 and https://github.com/Netflix/zuul/blob/abdc1bd416ccf1fd03c32a12a403ea8de0a84cf0/zuul-core/src/main/java/com/netflix/zuul/netty/ssl/BaseSslContextFactory.java#L164

**Comment by artgon** [comment]
I think maybe I miscommunicated the changes. All the code you had before was fine, just wanted the creation method in the initialized to move to a class `ClientSslContextFactory` that extends `BaseSslContextFactory`. There's no need to create a new interface or copy code to conn pool config. You can reference those static methods I linked to directly, and use the inherited methods directly in your subclass.

**Comment by artgon** [comment]
Let's go back to the original commit, get it merged and we can improve it iteratively.

**Comment by JacobJohansen** [comment]
The rough part in using the baseSslContextFactory is that its tightly coupled to the server configuration. This was my first attempt at separating the server config from an actual base ssl context provider. 

I’ll revert it back to the simpler version when I'm back at my desktop.

**Comment by JacobJohansen** [comment]
@artgon I was able to get in the changes for ciphers and protocols in a simpler way. hope this is the best of both worlds.

**Comment by artgon** [comment]
I think that couple between the `SslConfigs` and `ConnectionPoolImpl` is unnecessary. I agree that those classes can be refactored to be better decoupled but I don't think you need to add it as part of this PR. That's why I wanted the initial, simple implementation you had.

**Comment by artgon** [comment]
This is great. Thanks for your contribution!

**Pull request merged**